### PR TITLE
VIB-50: Create Project model for storing synced data from Google Sheet or Exce

### PR DESCRIPTION
**Pull request by PivtoranisV**
[VIB-50](https://cloverpop-internship-2025.atlassian.net/browse/VIB-50) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=wahanegi&utm_campaign=badge_info)

Hello team, please review PR

## Descriptions
- Add project model
- Add specs for the project model
- Add Active Admin for Project model

Active Admin Projects window:
Screenshot 2025-02-11 at 12 09 22

### CODE REVIEW

**Review by rogergraves — COMMENTED** [review]
It came to mind that in order to avoid situations where there might be duplications introduced in the `code` field due to uppercase vs. lowercase (for example `123-kpp-321` versus `123-KPP-321` what do you think about adding an `before_save` that sets the field to `.upcase`? For example:

```ruby
before_save :upcase_code

validates :code, presence: true, uniqueness: { case_sensitive: false }

def upcase_code
   self.code = code.upcase if code.present?
end
```

File: db/migrate/20250207121456_create_projects.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateProjects < ActiveRecord::Migration[7.1]
+  def change
+    create_table :projects do |t|
+      t.string :company
+      t.string :code
+      t.string :name
```
Comment (Maryna-Harasko) [inline db/migrate/20250207121456_create_projects.rb:6]: It would be better to add null: false to the company, code, and name fields in the migration to prevent NULL values in the database and ensure consistency with the model validations.

File: db/migrate/20250211134135_add_not_null_constraints_to_projects.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class AddNotNullConstraintsToProjects < ActiveRecord::Migration[7.2]
+  def change
+    change_table :projects, bulk: true do |t|
+      t.change_null :code, false
+      t.change_null :company, false
+      t.change_null :name, false
+    end
+  end
+end
```
Comment (Maryna-Harasko) [inline db/migrate/20250211134135_add_not_null_constraints_to_projects.rb:9]: Since the migration has only been applied in the current branch, it is recommended to perform a rollback and reapply the migration with the necessary changes to ensure data integrity.

Comment (rogergraves) [reply db/migrate/20250211134135_add_not_null_constraints_to_projects.rb:9]: @Maryna-Harasko @PivtoranisV I think it would be better to enforce these things within Rails instead of the database.

Comment (PivtoranisV) [reply db/migrate/20250211134135_add_not_null_constraints_to_projects.rb]: Revert changes. code, company and project validation for presence is done only on the Rails side
Thank you

**Review by Maryna-Harasko — APPROVED** [review]

File: spec/factories/projects.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#  updated_at :datetime         not null
+#
+# Indexes
+#
+#  index_projects_on_code  (code) UNIQUE
+#
+FactoryBot.define do
+  factory :project do
+    company { Faker::Company.name }
+    code { Faker::Alphanumeric.alphanumeric(number: 8) }
```
Comment (rogergraves) [inline spec/factories/projects.rb:19]: Please specify to Faker that this field should be unique

File: spec/models/project_spec.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    end
+  end
+
+  context 'Validations' do
+    describe 'company' do
+      it { is_expected.to validate_presence_of(:company) }
+      it { is_expected.to allow_value('Tech Corp').for(:company) }
+    end
+
+    describe 'code' do
```
Comment (rogergraves) [inline spec/models/project_spec.rb:33]: Please add more testing for the validations that you have by addressing edge cases, for example:

```ruby
let(:project1) { FactoryBot.create(:project, code: '123-JJJ' }
let(:project2) { FactoryBot.create(:project, code: 'JJJ-321') }

it 'codes that use lower case letters and spaces still have uniqueness enforced on them' do
   project2.code = ' 123-Jjj '
   expect(project2.valid?).to be_falsey 
   expect(project2.save).to be_falsey
   expect(project2.code).to eq('JJJ-321')
end
```
If this fails you may want to adjust the validations in the model until it passes.

**Review by rogergraves — CHANGES_REQUESTED** [review]

File: spec/factories/projects.rb (hunk 3 above)
Comment (PivtoranisV) [reply spec/factories/projects.rb]: Thank you, Roger. Update Faker to genrate unique codes

File: spec/models/project_spec.rb (hunk 4 above)
Comment (PivtoranisV) [reply spec/models/project_spec.rb:33]: Hi Roger, thank you for that. 
Yes, you are right, these edge cases are failing. I updated the project model to make it pass. Instead of before_save callback, I found that in Rails 7.1, the normalize method was introduced. It works for these edge cases, also I tested on ActiveAdmin these edge cases, and all work as expected
https://api.rubyonrails.org/v7.2/classes/ActiveRecord/Normalization/ClassMethods.html

**Review by rogergraves — APPROVED** [review]
